server/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -417,81 +467,39 @@ private void loadAllLookupsAndInitStateRef()
+    CompletionService<Map.Entry<String, LookupExtractorFactoryContainer>> completionService =
+        new ExecutorCompletionService<>(executorService);
+    List<Future<?>> allFutures = new ArrayList<>();
     try {
       LOG.info("Starting lookup loading process");
-      for (LookupBean lookupBean : lookupBeanList) {
-        futures.add(
-            completionService.submit(
-                () -> {
-                  LookupExtractorFactoryContainer container = lookupBean.getContainer();
-                  LOG.info(
-                      "Starting lookup [%s]:[%s]",
-                      lookupBean.getName(),
-                      container
-                  );
-                  if (container.getLookupExtractorFactory().start()) {
-                    LOG.info(
-                        "Started lookup [%s]:[%s]",
-                        lookupBean.getName(),
-                        container
-                    );
-                    return new AbstractMap.SimpleImmutableEntry<>(lookupBean.getName(), container);
-                  } else {
-                    LOG.error(
-                        "Failed to start lookup [%s]:[%s]",
-                        lookupBean.getName(),
-                        container
-                    );
-                    return null;
-                  }
-                }
-            )
-        );
-      }
-      for (int i = 0; i < futures.size(); i++) {
-        try {
-          final Future<AbstractMap.SimpleImmutableEntry<String, LookupExtractorFactoryContainer>> completedFuture = completionService
-              .take();
-          final AbstractMap.SimpleImmutableEntry<String, LookupExtractorFactoryContainer> lookupResult = completedFuture
-              .get();
-          if (lookupResult != null) {
-            builder.put(lookupResult);
-          }
-        }
-        catch (ExecutionException e) {
-          LOG.error(e, "Execution error during lookup loading.");
+      for (int i = 0; i < lookupConfig.getLookupStartRetries(); i++) {
+        LOG.info("Round of attempts #%d, [%d] lookups", i + 1, lookupBeanList.size());
+        Map<String, LookupExtractorFactoryContainer> successfulLookups =
+            startLookups(lookupBeanList, completionService, allFutures);
+        builder.putAll(successfulLookups);
+        List<LookupBean> failedLookups = lookupBeanList
+            .stream()
+            .filter(l -> !successfulLookups.containsKey(l.getName()))
+            .collect(Collectors.toList());
+        if (failedLookups.isEmpty()) {
+          break;
+        } else {
+          // next round
+          lookupBeanList = failedLookups;
         }
       }
       stateRef.set(new LookupUpdateState(builder.build(), ImmutableList.of(), ImmutableList.of()));
     }
     catch (Exception e) {
       LOG.error(e, "Failed to finish lookup load process.");
-      for (Future future : futures) {
+      for (Future future : allFutures) {
```

Comment:
Since allFutures doesn't really matter, and the success of the cancel doesnt matter, and the futures are collected by the queue in the service, can you just shutdownNow the service in a finally block and not track the futures?

Reply:
Removed